### PR TITLE
added array size support for structures

### PR DESCRIPTION
**Pull request by bluec0re**
```
struct test {
  char name[256];
  int id
}
```
to
```
pf [256]zd name id
```
instead of
```
pf *zd name id
```

### CODE REVIEW

**Comment by radare** [comment]
isnt this conflicting with this:

```
| pf [4]w[7]i                                  Print an array of 4 words and then an array of 7 integers
```

can you add tests for this?

**Comment by bluec0re** [comment]
Why should it? The commits only change the conversation of type definitions to pf commands.

**Comment by bluec0re** [comment]
But after thinking, is it possible that the header parsing results into wrong output in terms of array declarations? It seems that any array declaration (<type> <name>[<amount>]) results into 
```
struct.<structname>.<name> = <type>*,<offset>,<amount>
```

Shouldn't it be

**Comment by Maijin** [comment]
Shouldn't it be what ?

**Comment by bluec0re** [comment]
sry comment was sent to early.

**Comment by radare** [comment]
ah ok! didnt read the code O:)

**Comment by XVilka** [comment]
Related to https://github.com/radare/radare2/issues/3115

File: libr/anal/types.c [hunk 1]
```diff
@@ -228,11 +229,21 @@ R_API char *r_anal_type_format(RAnal *anal, const char *t) {
...
+					elements = sdb_array_get_num (DB, var2, 2, NULL);
 					snprintf (var3, sizeof (var3), "type.%s", type);
 					tfmt = sdb_const_get (DB, var3, NULL);
 					if (tfmt) {
 						filter_type (type);
-						fmt = r_str_concat (fmt, tfmt);
+						if (elements > 0) {
+							fmt = r_str_concatf (fmt, "[%d]", elements);
+							if (*tfmt == '*') { // is not pointer but local array
+								fmt = r_str_concat (fmt, tfmt+1);
```
Comment (radare) [inline libr/anal/types.c]: add spaces around +

**Comment by Maijin** [comment]
Great that will fix #3115

**Comment by bluec0re** [comment]
Not actually. I'm realized that it will only try to give an workaround for it (+ a correct output in terms of arrays). I think the correct solution would be to generate correct types in the parsing routines.

**Comment by Maijin** [comment]
? This seems to fix the issue for me since it creates [256] when there is a [256] etc.

**Comment by radare** [comment]
you can fix that space thing in another commit soon or late :P merged

**Comment by bluec0re** [comment]
That's right, but the parsing routines generate pointer types instead of passing the base type. I've used a workaround by skipping the first character if it's signaling a pointer type.

**Comment by Maijin** [comment]
@bluec0re  Add a test please

**Comment by bluec0re** [comment]
Btw. getting more into this, I think that this test is wrong:
https://github.com/radare/radare2-regressions/blob/master/t/cmd_types#L376

It generates a file like this:
```
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00000000  1300 0000 6400 0000 0000 0000 0000 0000  ................
0x00000010  0000 0000 0000 0000 0000 0000 0000 0000  ................
0x00000020  0000 0000 0000 0000 0000 0000 0000 0000  ................
0x00000030  0000 0000 0000 0000 0000 0000 0000 0000  ................
0x00000040  0000 0000 0000 0000 0000 0000 0000 0000  ................
0x00000050  0000 0000 0000 0000 0000 0000 0000 0000  ................
0x00000060  0000 0000 4361 726c 6f73 0000 0000 0000  ....Carlos......
```

Where it should be more like this
```
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00000000  1300 0000 4361 726c 6f73 0000 0000 0000  ....Carlos......

```

Based on the struct
```
struct person { int age; char name[10]; }
```

Should I also change this test? (and mark it broken)

**Comment by Maijin** [comment]
Fix the test with the correct output and yes if it doesn't pass yes BROKEN=1

**Pull request merged**